### PR TITLE
[DONOTMERGE] (Pending 5169) OpenShift Volume Security Integration

### PR DESCRIPTION
**Pull request by pweil-**
This is based on @pmorie 's WIP branch https://github.com/openshift/origin/pull/5169.  The last two commits are what I'm working on.

Contains:

1.  SecurityContextConstraints api support
   1.  Includes a defaulter for SCCs without FSGroup/SupplementalGroups strategies so no upgrade is necessary
   1.  Defaulting is based on the value of the RunAsUser strategy
1.  Strategies for groups (fsgroup and sup groups use the same underlying strategy code but have their own definitions/types)
   1.  Both strategies allow namespace configuration by using MustRunAs with no ranges set
   1.  Both strategies will provide a default value of the first value in the first range configured
   1.  When using namespace configuration for the FSGroup strategy it will create a single range with a min/max of the first value in the supplemental group annotation. 
   1.  When using namespace configuration for the SupGroup strategy it will parse the entire annotation into ranges.
1.  provider implementation
   1.  new interface methods introduced to generate and validate a PSC 
   1.  support for fsgroup/sup group strategies
   1.  reuse of the selinux strategy validation on the PSC by mocking up a container to validate (generation never uses the container so no mock is necessary)
1.  admission control for PSC
   1.  generates and validates the PSC
   1.  uses the SecurityContext provider to determine the effective SC for the container prior to container generation/validation (provides us initial value defaulting and allows container overrides)
   1.  retrieval of the supplemental group annotation has fallback code to use the uid annotation if necessary for a better upgrade experience

@pmorie @swagiaal @liggitt @smarterclayton 

---
Remaining Work:
 - [ ] blacklisting `v1beta3` `PodSpecs` has rippled into a lot of objects now not getting fuzzed.  At the very least, we need an issue find a way to test `v1beta3 -> internal -> v1beta3` with fidelity.  See the ripples here: https://github.com/openshift/origin/pull/5079/files#diff-394377845c8dfd9eec363c059cdde354R379
 - [ ] UPSTREAM commit names need fixing.
 - [ ] issue to allow admission based on a union of available SCCs.  With multiple ranges and options, its really annoying to have an SCC that combines all of them.
 - [ ] issue to convert code to a `DeepCopy` of the pod.  This would help readability and its only going to happen on pod creation.  I'm not convinced that an in-memory copy of data we already have, just for creates of pods, will show up on profiling, but it be a big improvement for inspectability of this code.
 - [ ] issue for `reconcile-sccs` to help admins update correctly.

### CODE REVIEW

**Comment by pmorie** [comment]
@pweil- I think we should make ginkgo tests for origin that test the knobs for docker in conjunction with the policy.  Long-term we need to have that for everything, we can start with just the scenarios being added (IE, SupplementalGroups and FSGroup).  Thoughts?

**Comment by pweil-** [comment]
are you proposing carrying upstream ginko tests that work with our carried scc or just adding them to security policy when it goes in the exp api?

**Comment by pmorie** [comment]
@pweil- Short term, i'm proposing e2es for origin that are probably similar to the ones for the knobs we make upstream that work in concert w/ the origin policy but don't share code with the upstream E2Es.  Ideally we would be able to intercept and decorate namespace creation in origin to set up policy but I don't want to have to work through that to solve the problem of just adding E2E for this feature in origin.

**Comment by pweil-** [comment]
latest update has supplemental groups in providers and strategies along with tests (still just the last two commits)

**Comment by pweil-** [comment]
provider using restricted scc to default the supplemental group to the first value of the allowable range

```
"securityContext": {
            "supplementalGroups": [
                1000010000
            ]
        },

"GroupAdd": [
            "1000010000"
        ],
```

**Comment by pweil-** [comment]
@smarterclayton @pmorie @swagiaal @liggitt - fsgroup integration is in for SCC.  The last two commits in this branch are what needs eyes, the rest are changes from @pmorie and @swagiaal that will be reviewed and submitted separately. 

```
"securityContext": {
            "supplementalGroups": [
                1000010000
            ],
            "fsGroup": 1000010000
        },
```

**Comment by pweil-** [comment]
@swagiaal  @pmorie  - a couple of things I'm noticing:

1.  There is nothing accounting for the new PodSecurityContext in the OpenShift v1beta3 code so round tripping a v1beta3 object would lose the PSC (make check is failing because the new fields aren't retained)
1.  swagger needs to be updated, when I verify on my branch it's showing changes for some of these commits
1.  gofmt on the commits is failing

**Comment by pweil-** [comment]
took care of the feedback for making constant field names in the provider.  Awaiting other reviews.  

reviewers: please pay attention to the conversions.  Since the upstream has diverged (it is now PSC) there is not a nice way to auto generate these so a double check would be good.

**Comment by smarterclayton** [comment]
Add one integration test that validates that the default behavior is to allocate the group id into the container

**Comment by pweil-** [comment]
added v1/v1beta3 defaults if new group strategy types are empty. Defaults will be for a restricted, namespace allocated strategy (mustRunAs with no ranges set).  

added integration test that submits pods with and without requested groups and ensures the groups are allocated if necessary and that they end up on the docker host config.

**Comment by pweil-** [comment]
updated to latest from @pmorie.  This now contains all required pieces.  Related PR that this depends on https://github.com/openshift/origin/pull/5169.

**Comment by pweil-** [comment]
@liggitt @pmorie @smarterclayton @deads2k - please put this in your review queue for the last 2 commits.  They are ready to wrap up.  Summary is in the main comment.

@deads2k - question on the extensions conversions/deep copies.  I notice in there that there is not a conversion/copy for things like Node and Pod.  But there is a conversion for PodSpec.  Do you know why?  I did not add conversions for SCC to extensions but I have them in another PR if they are necessary.

**Comment by pweil-** [comment]
@soltysh too for reviews

**Comment by pweil-** [comment]
Bump.  Monday mornings are perfect for SCC reviews.  Best to do when you wake up angry :smile:

**Comment by deads2k** [comment]
> @deads2k - question on the extensions conversions/deep copies. I notice in there that there is not a conversion/copy for things like Node and Pod. But there is a conversion for PodSpec. Do you know why? I did not add conversions for SCC to extensions but I have them in another PR if they are necessary.

the `PodSpec` was transitively required through the `DaemonSetSpec`.  If your object lives under an object that has conversions here, you probably need it.  The really scary question to ask yourself is why they have generated the same conversion function twice?  I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

**Comment by pweil-** [comment]
> the PodSpec was transitively required through the DaemonSetSpec. If your object lives under an object that has conversions here, you probably need it. The really scary question to ask yourself is why they have generated the same conversion function twice? I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

Ok, should be safe then as SCC is a standalone object.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go [hunk 1]
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
...
-		seLinuxStrategy:   seLinuxStrat,
+		scc:                       scc,
+		runAsUserStrategy:         userStrat,
+		seLinuxStrategy:           seLinuxStrat,
+		fsGroupStrategy:           fsGroupStrat,
+		supplementalGroupStrategy: supGroupStrat,
 	}, nil
 }
 
+// Create a PodSecurityContext based on the given constraints.  If a settingis already set
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: setting is

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go [hunk 2]
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
...
+// on the PodSecurityContext it will not be changed.  Validate should be used after the context
+// is created to ensureit complies with the required restrictions.
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: ensure it

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go [hunk 3]
```diff
@@ -41,45 +51,81 @@ func NewSimpleProvider(scc *api.SecurityContextConstraints) (SecurityContextCons
...
+	if len(sc.SupplementalGroups) == 0 {
+		supGroups, err := s.supplementalGroupStrategy.Generate(pod)
+		if err != nil {
+			return nil, err
+		}
+		sc.SupplementalGroups = supGroups
+	}
+
+	if sc.FSGroup == nil {
+		fsGroup, err := s.fsGroupStrategy.Generate(pod)
```
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: Why does this return a list if we're only using the first value?

Comment (pweil-) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: I want to be able to reuse the same scc strategies for sup group and fsgroup.  This allows custom strategies to provide ranges of acceptable groups and multiple defaults.  I thought it best to leave the interface flexible enough to provide multiple values.

Two things to note:

1.  fsgroup is a single field (so to reuse I need to make sure I only take a single value)
2.  the default mustRunAs generation is just returning the first value in the configured ranges anyway (but as mentioned, I thought it best to leave the interface capable of returning multiple to support defaulting to multiple sup groups)

Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: > the default mustRunAs generation is just returning the first value in the configured ranges anyway (but as mentioned, I thought it best to leave the interface capable of returning multiple to support defaulting to multiple sup groups)

Docs on the interface usage describe this behavior?  I think you shoudl still have two different interfaces, even if they overlap.

Comment (pweil-) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/securitycontextconstraints/provider.go]: I can live with that.  I'll add a generate single method.

File: pkg/api/serialization_test.go [hunk 4]
```diff
@@ -375,7 +375,9 @@ func roundTrip(t *testing.T, codec runtime.Codec, originalItem runtime.Object) {
 var skipStandardVersions = map[string][]string{
 	// The API versions here are to test our object that serializes from/into
 	// docker's registry API.
-	"DockerImage": {"pre012", "1.0"},
+	"DockerImage":          {"pre012", "1.0"},
+	"DeploymentConfigList": {"", "v1"},
```
Comment (deads2k) [inline pkg/api/serialization_test.go]: This is because we're breaking compatibility with `v1beta3`?  Where are the special cased tests to make sure I can still correctly parse these from `v1beta3` in?

Comment (pweil-) [inline pkg/api/serialization_test.go]: @pmorie :arrow_up:

File: UPGRADE.md [hunk 5]
```diff
@@ -32,8 +32,26 @@ At that time, the openshift docker registry image must be upgraded in order to c
...
+
+  1.  allowHostPID - defaults to false.  You may wish to change this to true on any privileged SCCs or 
+  [reset your default SCCs](https://docs.openshift.org/latest/admin_guide/manage_scc.html#updating-the-default-security-context-constraints) 
+  which will set this field to true for the privileged SCC and false for the restricted SCC.
+  1.  allowHostIPC - defaults to false.  You may wish to change this to true on any privileged SCCs or 
+  [reset your default SCCs](https://docs.openshift.org/latest/admin_guide/manage_scc.html#updating-the-default-security-context-constraints) 
+  which will set this field to true for the privileged SCC and false for the restricted SCC.
+  1.  fsGroup - if the strategy type is unset this field will default based on the runAsUser strategy.
+  If runAsUser is set to RunAsAny this field will also be set to RunAsAny.  If the strategy type is
+  any other value this field will default to MustRunAs and look to the namespace for annotation
```
Comment (deads2k) [inline UPGRADE.md]: And when that annotation is missing, what will it do?

File: pkg/api/serialization_test.go [hunk 6]
```diff
@@ -385,7 +387,7 @@ func TestSpecificKind(t *testing.T) {
 	api.Scheme.Log(t)
 	defer api.Scheme.Log(nil)
 
-	kind := "ImageStreamTag"
+	kind := "BuildConfig"
```
Comment (deads2k) [inline pkg/api/serialization_test.go]: This is a fun record of who had trouble with what most recently :)

File: pkg/security/util.go [hunk 7]
```diff
@@ -3,8 +3,7 @@ package security
 const (
 	UIDRangeAnnotation = "openshift.io/sa.scc.uid-range"
 	// SupplementalGroupsAnnotation contains a comma delimited list of allocated supplemental groups
-	// for the namespace.  Groups are in the form of individual group ids or a range of group ids.
-	// Range format is supported in the form of a Block which supports {start}/{length} or {start}-{end}
+	// for the namespace.  Groups are in the form of a Block which supports {start}/{length} or {start}-{end}
 	SupplementalGroupsAnnotation = "openshift.io/sa.scc.supplemental-groups"
```
Comment (deads2k) [inline pkg/security/util.go]: `sa`?

File: pkg/security/admission/admission.go [hunk 8]
```diff
@@ -177,11 +196,13 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
...
+		// ensure psc is not mutated if there are errors
+		pod.Spec.SecurityContext = originalPSC
 		return errs
 	}
 
 	// if we've reached this code then we've generated and validated an SC for every container in the
-	// pod so let's apply what we generated
+	// pod so let's apply what we generated.  Note: the psc is already applied.
 	for i, sc := range generatedSCs {
 		pod.Spec.Containers[i].SecurityContext = sc
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Haven't I already done this here: https://github.com/openshift/origin/pull/5079/files#diff-eb43ba779f411e3ee6b772cceee9810dR194

File: pkg/security/admission/admission.go [hunk 9]
```diff
@@ -177,11 +196,13 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
...
 	if len(errs) > 0 {
+		// ensure psc is not mutated if there are errors
+		pod.Spec.SecurityContext = originalPSC
 		return errs
 	}
 
 	// if we've reached this code then we've generated and validated an SC for every container in the
-	// pod so let's apply what we generated
+	// pod so let's apply what we generated.  Note: the psc is already applied.
 	for i, sc := range generatedSCs {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Why don't I just iterate over the `pod.Spec.Containers`?  That would be more obvious to me as a reader.

File: pkg/security/admission/admission.go [hunk 10]
```diff
@@ -164,7 +165,25 @@ func assignSecurityContext(provider scc.SecurityContextConstraintsProvider, pod
...
+	}
+
+	// save the original PSC and validate the generated PSC.  Leave the generated PSC
+	// set for container generation/validation.  We will reset to original post container
+	// validation.
+	originalPSC := pod.Spec.SecurityContext
+	pod.Spec.SecurityContext = psc
+	errs = append(errs, provider.ValidatePodSecurityContext(pod).Prefix("spec.securityContext")...)
+
 	for i, c := range pod.Spec.Containers {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Ok, looking through this loop, its really hard to read what's modifying a container and what is modifying a copy of a container that is going to be discarded later.

Re-reading, I **think** that this use a pointer to a copy of a container to perform a bunch of mutations that are ultimately discarded.  If that's right, can you rename `c` to `containerCopy` or some such?

File: pkg/security/admission/admission.go (hunk 9 above)
Comment (deads2k) [inline pkg/security/admission/admission.go]: > Why don't I just iterate over the pod.Spec.Containers? That would be more obvious to me as a reader.

I think I see now.

File: pkg/security/admission/admission.go [hunk 11]
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 				continue
 			}
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Why can't I close the `if requiresNamespaceAllocations` here?

File: pkg/security/util.go (hunk 7 above)
Comment (pweil-) [inline pkg/security/util.go]: Yeah, I kept the new name consistent with the old names which have sa in there.

File: pkg/security/admission/admission.go [hunk 12]
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
...
 
+			// Make a copy of the constraint so we don't mutate the store's cache
+			var constraintCopy kapi.SecurityContextConstraints = *constraint
+			constraint = &constraintCopy
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: This is warping my mind a little.  Can we move it to the top of the loop, choose a non-conflicting name, and assign the final thing to `constraintCopy`?  Something like this:

```go
for _, scc := range sccs{
    var t kapi.SecurityContextConstraints = *scc
    sccCopy := &t
```

The name change would ensure you don't accidentally leave any stragglers.

File: pkg/security/admission/admission.go [hunk 13]
```diff
@@ -205,8 +226,11 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 		var err error
 		resolveUIDRange := requiresPreAllocatedUIDRange(constraint)
 		resolveSELinuxLevel := requiresPreAllocatedSELinuxLevel(constraint)
+		resolveFSGroup := requiresPreallocatedFSGroup(constraint)
+		resolveSupplementalGroups := requiresPreallocatedSupplementalGroups(constraint)
+		requiresNamespaceAllocations := resolveUIDRange || resolveSELinuxLevel || resolveFSGroup || resolveSupplementalGroups
 
-		if resolveUIDRange || resolveSELinuxLevel {
+		if requiresNamespaceAllocations {
 			var min, max *int64
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: seems like these should be scoped to the blocks where you need them, so that they never escape.

File: pkg/security/admission/admission.go [hunk 14]
```diff
@@ -216,40 +240,51 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 				continue
 			}
 
+			// Make a copy of the constraint so we don't mutate the store's cache
+			var constraintCopy kapi.SecurityContextConstraints = *constraint
+			constraint = &constraintCopy
+
 			// Resolve the values from the namespace
 			if resolveUIDRange {
 				if min, max, err = getPreallocatedUIDRange(namespace); err != nil {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: shouldn't this be able to write as 
```go
constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace)
if err != nil{
    // do stuff
}
```

File: pkg/security/admission/admission.go (hunk 12 above)
Comment (deads2k) [inline pkg/security/admission/admission.go]: Why not make this a deep copy to avoid these shenanigans later on: https://github.com/openshift/origin/pull/5079/files#diff-eb43ba779f411e3ee6b772cceee9810dR262

File: pkg/security/admission/admission.go [hunk 15]
```diff
@@ -349,6 +384,91 @@ func getPreallocatedLevel(ns *kapi.Namespace) (string, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	if len(groups) == 0 {
+		return nil, fmt.Errorf("found annotation %s but it was empty", allocator.SupplementalGroupsAnnotation)
+	}
+	glog.V(4).Infof("got preallocated value for groups: %s in namespace %s", groups, ns.Name)
+
+	blocks, err := parseSupplementalGroupAnnotation(groups)
+	if err != nil {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Why aren't I allowed to have zero supplemental groups?

**Comment by deads2k** [comment]
I found some of the code hard to parse.  After its updated, I'll go back through and check test coverage.   I'm assuming I'll find unit tests for the different states of SCC, namespace annotations, and constraints.  I think the tests will be easier to read if they focus on one contraint at a time instead of one giant method testing all of them.

File: pkg/security/util.go (hunk 7 above)
Comment (deads2k) [inline pkg/security/util.go]: > Yeah, I kept the new name consistent with the old names which have sa in there.

Ok.  For the historical record, what does that stand for?

File: pkg/security/admission/admission.go (hunk 11 above)
Comment (pweil-) [inline pkg/security/admission/admission.go]: resolving the ns values is only applicable to something that needs ns allocation.  It could be closed since the other flags gate resolving individual values if that helps readability.

File: pkg/security/util.go (hunk 7 above)
Comment (pweil-) [inline pkg/security/util.go]: It was originally meant to stand for service account but is not applicable since an SA isn't directly assigned these values, the ns is.  I will remove it from the newest annotation to not perpetuate wrongness for the sake of consistency unless you disagree with that approach and would rather keep it consistent.

File: pkg/security/admission/admission.go (hunk 11 above)
Comment (deads2k) [inline pkg/security/admission/admission.go]: > resolving the ns values is only applicable to something that needs ns allocation. It could be closed since the other flags gate resolving individual values if that helps readability.

Yes, close here.  It helps break this apart into congruent chunks with a precondition.

File: pkg/security/admission/admission.go [hunk 16]
```diff
@@ -205,8 +226,11 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
 		var err error
 		resolveUIDRange := requiresPreAllocatedUIDRange(constraint)
 		resolveSELinuxLevel := requiresPreAllocatedSELinuxLevel(constraint)
+		resolveFSGroup := requiresPreallocatedFSGroup(constraint)
+		resolveSupplementalGroups := requiresPreallocatedSupplementalGroups(constraint)
+		requiresNamespaceAllocations := resolveUIDRange || resolveSELinuxLevel || resolveFSGroup || resolveSupplementalGroups
 
-		if resolveUIDRange || resolveSELinuxLevel {
+		if requiresNamespaceAllocations {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: Below (its not in the diff), don't create a scoped `if` that doesn't scope.

```go
if namespace, err = c.getNamespace(ns, namespace); err != nil {
```
should be 
```go
namespace, err = c.getNamespace(ns, namespace)
if err != nil {
```

The former tricks me into thinking that the variables are scoped and don't leak out, the latter makes it clear that those variables are not scoped to the block.

File: pkg/security/util.go (hunk 7 above)
Comment (deads2k) [inline pkg/security/util.go]: > It was originally meant to stand for service account but is not applicable since an SA isn't directly assigned these values, the ns is. I will remove it from the newest annotation to not perpetuate wrongness for the sake of consistency unless you disagree with that approach and would rather keep it consistent.

Go ahead and keep them consistent.

File: pkg/security/admission/admission.go (hunk 15 above)
Comment (pweil-) [inline pkg/security/admission/admission.go]: When using a strategy that requires values you'll get an error creating the strategy if no values are given.  This allows us to indicate that the error was in parsing the annotation rather than a generic "bad strategy" message.

**Comment by deads2k** [comment]
put v1beta3 code into a separate commit so we can squash it into the v1beta3 carry later on.

File: pkg/security/admission/admission.go (hunk 8 above)
Comment (pweil-) [inline pkg/security/admission/admission.go]: Think you worked this one out yourself below.  Making the naming change to the loop above so it is more obvious that we aren't changing the original container.

**Comment by pweil-** [comment]
@deads2k - updated for feedback.  Broke into a "normal" and "v1beta3" commit for upstreams.

**Comment by pweil-** [comment]
upgrade.md updated.  Current doc that the link is pointing to: https://github.com/openshift/openshift-docs/pull/1078/files#diff-ac08c0344bc91609b46aae5f51152c8d

**Comment by deads2k** [comment]
I'll go back through again tomorrow morning, but this is looking good.

File: pkg/security/admission/admission.go [hunk 17]
```diff
@@ -205,51 +228,62 @@ func (c *constraint) createProvidersFromConstraints(ns string, sccs []*kapi.Secu
...
 
-			// Resolve the values from the namespace
-			if resolveUIDRange {
-				if min, max, err = getPreallocatedUIDRange(namespace); err != nil {
-					errs = append(errs, fmt.Errorf("unable to find pre-allocated uid annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err))
-					continue
-				}
+		// Resolve the values from the namespace
+		if resolveUIDRange {
+			if constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace); err != nil {
```
Comment (deads2k) [inline pkg/security/admission/admission.go]: This isn't scoping the values.  It should be:
```go
constraint.RunAsUser.UIDRangeMin, constraint.RunAsUser.UIDRangeMax, err = getPreallocatedUIDRange(namespace)
if err != nil {
```

**Comment by pweil-** [comment]
[test]

**Comment by soltysh** [comment]
> the PodSpec was transitively required through the DaemonSetSpec. If your object lives under an object that has conversions here, you probably need it. The really scary question to ask yourself is why they have generated the same conversion function twice? I suspect it was done because they did not properly indicate which schemes/group/version a given group/version relies upon.

I've raised that issue very early when implementing jobs (see https://github.com/kubernetes/kubernetes/issues/12977).

**Comment by soltysh** [comment]
@pweil- I've left you some comments, mostly stylistic. Other than that it's looking good.

**Comment by deads2k** [comment]
The upstream commits need some renaming before merge.

**Comment by deads2k** [comment]
Also, looks like your new test isn't passing:
```
--- FAIL: TestSupplementalGroups-2 (13.37s)
	supplemental_groups_test.go:121: timed out waiting for the condition
FAIL
```

**Comment by pweil-** [comment]
@soltysh - comments addressed, please take another pass.   I left the defaulter based on @deads2k comment.

File: pkg/security/admission/admission_test.go [hunk 18]
```diff
@@ -137,12 +178,30 @@ func TestAdmit(t *testing.T) {
...
-		expectedLevel string
-		expectedPriv  bool
+		pod               *kapi.Pod
+		shouldAdmit       bool
+		expectedUID       int64
+		expectedLevel     string
+		expectedFSGroup   int64
+		expectedSupGroups []int64
+		expectedPriv      bool
+		podLevelOnly      bool
```
Comment (deads2k) [inline pkg/security/admission/admission_test.go]: godoc on this particular field.

File: pkg/security/admission/admission_test.go [hunk 19]
```diff
@@ -205,11 +305,31 @@ func TestAdmit(t *testing.T) {
 			if validatedSCC != saSCC.Name {
 				t.Errorf("%s should have validated against %s but found %s", k, saSCC.Name, validatedSCC)
 			}
-			if *v.pod.Spec.Containers[0].SecurityContext.RunAsUser != v.expectedUID {
-				t.Errorf("%s expected UID %d but found %d", k, v.expectedUID, *v.pod.Spec.Containers[0].SecurityContext.RunAsUser)
+
+			if !v.podLevelOnly {
+				// ensure anything we expected to be defaulted on the container level is set
+				if *v.pod.Spec.Containers[0].SecurityContext.RunAsUser != v.expectedUID {
```
Comment (deads2k) [inline pkg/security/admission/admission_test.go]: Why aren't you be able to predict what the expected values should be?

**Comment by deads2k** [comment]
Is there a reasonable spot to actually test that this works end to end?  I'm convinced that this code will make sure the Pod/Containers are marked as expected, but I'd really like a test that makes sure that the expected user/group actually works how we want.  That probably belongs in extended, but it would make me feel a lot better.

**Comment by deads2k** [comment]
I added a checklist to the description of the things I'd like addressed before merge.  Nothing too big.

Last three commits lgtm.

**Comment by pweil-** [comment]
> Is there a reasonable spot to actually test that this works end to end? I'm convinced that this code will make sure the Pod/Containers are marked as expected, but I'd really like a test that makes sure that the expected user/group actually works how we want. That probably belongs in extended, but it would make me feel a lot better.

The sup group integration test is attempting to ensure that it works all the way down to the docker host config of a launched container.  Support for the fields in docker is 1.8.2 with patches applied.  @pmorie and @swagiaal should be ensuring the docker and beyond portion in addition to the application to the host config so we should be doubly covered on some parts.

**Comment by pweil-** [comment]
@pmorie - seems to be failing the int test due to 

```
E1027 10:26:09.747152    1208 mount_linux.go:103] Mount failed: exit status 1
Mounting arguments: tmpfs /tmp/openshift-integration/volume/pods/aade3400-7cb6-11e5-aac7-0ecd0b4a7e19/volumes/kubernetes.io~secret/default-token-nk9tf tmpfs [rootcontext="unconfined_u:object_r:svirt_sandbox_file_t:s0"]
Output: mount: only root can use "--options" option
```
suggestions?

**Comment by liggitt** [comment]
you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

**Comment by pweil-** [comment]
> you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

is the router test somehow excluded then?  It starts a router pod.  I thought we used the docker build flag specifically to run things with docker.

**Comment by pweil-** [comment]
never mind, it starts it itself I think

**Comment by pweil-** [comment]
modified the int test to launch the pod but then not check for the docker host config.  It will now simulate what the kubelet does to craft the host config using the sc provider and checks the modified host config.

**Comment by deads2k** [comment]
> you can't run pods in integration... it doesn't start with sudo, so it can't use tmpfs

This should be fixable on jenkins, right?

File: test/integration/supplemental_groups_test.go [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+	defer kubeClient.Pods(testutil.Namespace()).Delete(pod.Name, nil)
+
+	// get the pod which should now be updated.  If admission mutated the pod unexpectedly
+	// then the checks for our submitted groups will fail (below)
+	updatedPod, err := kubeClient.Pods(testutil.Namespace()).Get(pod.Name)
+	if err != nil {
+		t.Fatalf("unable to get updated pod: %v", err)
+	}
+
+	// simulate what the kubelet would do to modify the container's host config
```
Comment (deads2k) [inline test/integration/supplemental_groups_test.go]: This leaves me unsatisfied.  I'd rather see an extended test get its pod created or fix the integration test invocation to allow running pods.

Comment (liggitt) [inline test/integration/supplemental_groups_test.go]: if tmpfs for the SA token is the issue, you can create a volume mount at the expected SA token mount path and the SA token won't get added.

Comment (pweil-) [inline test/integration/supplemental_groups_test.go]: I'll give that a shot.  Worse comes to worse I can move this to an extended test.

**Comment by pweil-** [comment]
[test][extended:security]

**Comment by pweil-** [comment]
@deads2k 

```
test/extended/core.sh --focus="security" -v

=== RUN "test/extended/security/supplemental_groups.go":
... snip ...
STEP: creating a pod that requests supplemental groups
STEP: retrieving the pod and ensuring groups are set
STEP: waiting for the pod to become running
STEP: finding the docker container id on the pod
STEP: creating a docker client and inspecting the container
STEP: ensuring the groups are set
STEP: Destroying namespace "e2e-tests-security-supgroups-qzl16" for this suite.
    ok       security: supplemental groups Ensure supplemental groups propagate to docker should propagate reques ... (9.17716644s)
```

**Comment by pweil-** [comment]
re[test][extended]

**Comment by pweil-** [comment]
let's try this again without the bad tag....re[test]

**Comment by pweil-** [comment]
@pmorie @deads2k @swagiaal - this didn't seem to work in jenkins although it worked fine locally (not run as root as well).  I'm going to add a bit more logging to the test to try and work it out.  Does anyone know the docker version that jenkins runs?  

@swagiaal I recall you saying previously that if the version was too low it would just ignore the parameters.

**Comment by deads2k** [comment]
> Does anyone know the docker version that jenkins runs?

@danmcp Do you know offhand?

**Comment by danmcp** [comment]
https://ci.openshift.redhat.com/jenkins/job/devenv_ami/lastStableBuild/consoleFull

and search for docker.

docker.x86_64                 1.7.1-115.el7

**Comment by pweil-** [comment]
@pmorie @deads2k I added a check on the extended test to check the version before running or skip.  Please take a glance.  New local output is showing the step for ensuring the feature is supported: https://gist.github.com/pweil-/9a783b6f4946e54e4933

**Comment by pweil-** [comment]
flake timeout on jenkins in unrelated tests.  re[test]

**Comment by swagiaal** [comment]
> @swagiaal I recall you saying previously that if the version was too low it would just ignore the parameters.

@pweil- that is correct

**Comment by pweil-** [comment]
it is correctly skipping the test now by checking the docker version:

```
=== RUN "test/extended/security/supplemental_groups.go":
>>> testContext.KubeConfig: /tmp/openshift-extended-tests/core/openshift.local.config/master/admin.kubeconfig
    skip    security: supplemental groups Ensure supplemental groups propagate to docker should propagate reques ... (7.086870098s)
```

It failed on unrelated extended tests.

@pmorie if you want to double check the last three commits to make sure you have the latest I'm fine with closing this out in favor of your updated PR.

Final status here:  

1.  LGTM from @deads2k on the last 3 commits
1.  New extended test will skip unsupported docker versions (clean run output above)
1.  Checklist on the top comment should migrate to @pmorie's PR.  I will create follow ups and he will address other items

**Comment by soltysh** [comment]
Add to that my LGTM :+1:

**Comment by pweil-** [comment]
Closing in favor of https://github.com/openshift/origin/pull/5169 :birthday:
